examples/with-apollo/lib/withData.js

```diff
@@ -0,0 +1,44 @@
+import { ApolloProvider, getDataFromTree } from 'react-apollo'
+import React from 'react'
+import 'isomorphic-fetch'
+import { IS_SERVER } from './exenv'
+import { initClient } from './initClient'
+import { initStore } from './initStore'
+
+export default (Component) => (
+  class extends React.Component {
+    static async getInitialProps (ctx) {
+      const headers = ctx.req ? ctx.req.headers : {}
+      const client = initClient(headers)
+      const store = initStore(client, client.initialState)
+
+      if (IS_SERVER) {
+        const app = (
+          <ApolloProvider client={client} store={store}>
+            <Component url={{ query: ctx.query, pathname: ctx.pathname }} />
+          </ApolloProvider>
+        )
+        await getDataFromTree(app)
+      }
+
+      return {
+        initialState: store.getState(),
```

Comment:
it only explodes with required query variables which are not met because the wrong query is associated (pre-populated query store and apollo client core starting at 1 again on the client). So the order of your components plays a role in triggering the bug – super spooky, hard to find and debug.